src/ast/passes/codegen_llvm.cpp

```diff
@@ -3006,98 +3052,109 @@ void CodegenLLVM::visit(Probe &probe)
+                     ? b_.CreateMapKeyAllocation(key_type,
+                                                 map.ident + "_key",
+                                                 key_expr->loc)
+                     : scoped_key_expr.value();
     if (inBpfMemory(key_expr->type)) {
       if (!key_expr->type.IsSameSizeRecursive(key_type)) {
         b_.CreateMemsetBPF(key, b_.getInt8(0), key_type.GetSize());
         if (key_expr->type.IsTupleTy()) {
-          createTupleCopy(key_expr->type, key_type, key, expr_);
+          createTupleCopy(
+              key_expr->type, key_type, key, scoped_key_expr.value());
         } else if (key_expr->type.IsStringTy()) {
-          b_.CreateMemcpyBPF(key, expr_, key_expr->type.GetSize());
+          b_.CreateMemcpyBPF(key,
+                             scoped_key_expr.value(),
+                             key_expr->type.GetSize());
         } else {
           LOG(BUG) << "Type size mismatch. Key Type Size: "
                    << key_type.GetSize()
                    << " Expression Type Size: " << key_expr->type.GetSize();
         }
       } else {
         // Call-ee freed
-        scoped_del.disarm();
       }
     } else if (map.key_type.IsIntTy()) {
       // Integers are always stored as 64-bit in map keys
-      b_.CreateStore(
-          b_.CreateIntCast(expr_, b_.getInt64Ty(), key_expr->type.IsSigned()),
-          key);
+      b_.CreateStore(b_.CreateIntCast(scoped_key_expr.value(),
+                                      b_.getInt64Ty(),
+                                      key_expr->type.IsSigned()),
+                     key);
     } else {
       if (key_expr->type.IsArrayTy() || key_expr->type.IsRecordTy()) {
         // We need to read the entire array/struct and save it
         b_.CreateProbeRead(ctx_,
                            b_.CreatePointerCast(
                                key, b_.GetType(key_expr->type)->getPointerTo()),
                            key_expr->type,
-                           expr_,
+                           scoped_key_expr.value(),
                            key_expr->loc);
       } else {
-        b_.CreateStore(
-            b_.CreateIntCast(expr_, b_.getInt64Ty(), key_expr->type.IsSigned()),
-            b_.CreatePointerCast(key, expr_->getType()->getPointerTo()));
+        b_.CreateStore(b_.CreateIntCast(scoped_key_expr.value(),
+                                        b_.getInt64Ty(),
+                                        key_expr->type.IsSigned()),
+                       b_.CreatePointerCast(
+                           key,
+                           scoped_key_expr.value()->getType()->getPointerTo()));
       }
     }
+    // Either way we hold on to the original key, to ensure that its lifetime
+    // lasts as long as it may be accessed.
+    return ScopedExpr(
+        key,
+        [this, key, k = std::move(scoped_key_expr), alloca_created_here]() {
+          if (alloca_created_here && dyn_cast<AllocaInst>(key))
```

Comment:
can we conditionally create this deleter based on `alloca_created_here` and `dyn_cast<AllocaInst>(key)`?

Reply:
Done.